### ISSUE TITLE
RFC: Consider updating `copy` semantics in `astype` to `False/None/True`

### ISSUE THREAD

**Issue by Micky774** [data-apis/array-api#788]
Currently the specification for `astype`, as added in https://github.com/data-apis/array-api/pull/290, specifies that `copy=False/True` wherein there is no room for a "copy never" option, and the default `copy=True` means that calls to `astype` that do not specify the `copy` kwarg never have a chance to be a no-op -- this results in unnecessary copies as _default_ behavior.

Other functions which use the `copy=False/None/True` semantics include:
- `asarray`
- `__dlpack__`
- `from_dlpack`
- `reshape`

As far as I can tell, `astype` is the exception to the rule.

**Comment by kgryte** [comment]
The default specification behavior follows NumPy and its derivatives. From the [NumPy docs](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html),

> By default, astype always returns a newly allocated array.

and the signature

```
ndarray.astype(dtype, order='K', casting='unsafe', subok=True, copy=True)
```

The default is `True`, not `None`. So the original specification simply followed suit.

**Comment by jakevdp** [comment]
If NumPy is the motivation for this, then I think there may be a problem: NumPy has explicitly decided to differ from the Array API in this respect; see https://numpy.org/neps/nep-0056-array-api-main-namespace.html#copy-keyword-semantics

**Comment by kgryte** [comment]
My reading from that NEP is not that NumPy is choosing to differ. NumPy is following the spec where `copy` kwarg behavior for `astype` is not the same as for `asarray` and others.

However, the situation is actually reversed. We based the current spec on what NumPy and its kin did at the time, which was to, by default, always copy.

**Comment by jakevdp** [comment]
I see, thanks.

**Comment by jakevdp** [comment]
In NumPy's implementation, do `copy=None` and `copy=False` have the same semantics of "copy only if necessary"?

**Comment by rgommers** [comment]
> In NumPy's implementation, do `copy=None` and `copy=False` have the same semantics of "copy only if necessary"?

Yes. But that's only because there was resistance to changing `copy=False` behavior to match the `asarray` semantics, and it didn't seem nearly as important as `asarray`. So `astype(..., copy=False)` still means "if needed" not "never".

**Comment by rgommers** [comment]
Maybe it'll be changed on a longer timescale, I don't know.

**Comment by Micky774** [comment]
Although the historical creation of the `astype` copy semantics makes sense (from both NumPy and the array API) I personally think at this point the standardization of other copy kwargs to include `None` leaves the API for `astype` confusing/misleading for developers that aren't quite as familiar with it. I still think it is worth changin

**Comment by asmeurer** [comment]
I think you can get "copy never" behavior in the current standard using `asarray(copy=None)` instead of `astype`. That's mostly equivalent to `if x.dtype == new_dtype` but I guess there's also subtleties with the devices.

**Comment by rgommers** [comment]
> `copy=True` is imo a bad default that leads to more memory movement than may be necessary

I am trying to figure out why this matters for JAX, since it doesn't actually have to do a physical memory copy - by design it guarantees that there is no difference (immutability). My understanding is that JAX had a `copy` keyword already in several places to match NumPy's signatures, that it didn't actually make physical memory copies when that was implemented, and that nothing has to change because of the array API standard. The addition of a true "never copy" mode was a lot of work in NumPy (and will be in CuPy as well), but should not have impacted JAX I believe.

This has now come up several times, so we really should make this more clear. The first time was at https://github.com/data-apis/array-api/issues/495#issuecomment-1286168414 I believe. And more recently, we had a more extensive discussion on this in https://github.com/dmlc/dlpack/pull/136 for DLPack. For DLPack it's about exchange between two libraries rather than semantics with a single library, so we put more effort into the "what does _copy_ actually mean".  

In summary, the `copy` keyword is a "logical copy", i.e. the returned data is guaranteed to be unique for all intents and purposes. If that can be achieved without moving data in memory, then that is fine. This also aligns with https://data-apis.org/array-api/draft/purpose_and_scope.html, which described some of the design principles of the standard - in particular that we want to describe semantics of functions without prescribing _execution_ semantics to the extent possible.

**Comment by jakevdp** [comment]
> I am trying to figure out why this matters for JAX

True copies are sometimes important in JAX; for example functions can be called with donated buffers, and donated buffers cannot be used in subsequent function calls. If a user wants a copy of an array for this purpose, we currently recommend `jnp.array(x, copy=True)` which copies data from the original array into a new buffer.

**Comment by crusaderky** [comment]
xref https://github.com/data-apis/array-api/issues/867
